File: src/main/java/net/pterodactylus/sone/web/SoneTemplatePage.java
/*
 * Sone - SoneTemplatePage.java - Copyright © 2010–2016 David Roden
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package net.pterodactylus.sone.web;

import java.io.UnsupportedEncodingException;
import java.net.URLEncoder;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collection;
import java.util.Collections;
import java.util.List;
import java.util.Map;

import javax.annotation.Nonnull;

import net.pterodactylus.sone.data.Sone;
import net.pterodactylus.sone.main.SonePlugin;
import net.pterodactylus.sone.web.page.FreenetRequest;
import net.pterodactylus.sone.web.page.FreenetTemplatePage;
import net.pterodactylus.util.notify.Notification;
import net.pterodactylus.util.template.Template;
import net.pterodactylus.util.template.TemplateContext;

import freenet.clients.http.SessionManager.Session;
import freenet.clients.http.ToadletContext;
import freenet.support.api.HTTPRequest;

import com.google.common.collect.ImmutableList;
import com.google.common.collect.ImmutableMap;

/**
 * Base page for the Sone web interface.
 *
 * @author <a href="mailto:bombe@pterodactylus.net">David ‘Bombe’ Roden</a>
 */
public class SoneTemplatePage extends FreenetTemplatePage {

	/** The Sone core. */
	protected final WebInterface webInterface;

	/** The page title l10n key. */
	private final String pageTitleKey;

	/** Whether to require a login. */
	private final boolean requireLogin;

	/**
	 * Creates a new template page for Sone that does not require the user to be
	 * logged in.
	 *
	 * @param path
	 *            The path of the page
	 * @param template
	 *            The template to render
	 * @param pageTitleKey
	 *            The l10n key of the page title
	 * @param webInterface
	 *            The Sone web interface
	 */
	public SoneTemplatePage(String path, Template template, String pageTitleKey, WebInterface webInterface) {
		this(path, template, pageTitleKey, webInterface, false);
	}

	/**
	 * Creates a new template page for Sone.
	 *
	 * @param path
	 *            The path of the page
	 * @param template
	 *            The template to render
	 * @param webInterface
	 *            The Sone web interface
	 * @param requireLogin
	 *            Whether this page requires a login
	 */
	public SoneTemplatePage(String path, Template template, WebInterface webInterface, boolean requireLogin) {
		this(path, template, null, webInterface, requireLogin);
	}

	/**
	 * Creates a new template page for Sone.
	 *
	 * @param path
	 *            The path of the page
	 * @param template
	 *            The template to render
	 * @param pageTitleKey
	 *            The l10n key of the page title
	 * @param webInterface
	 *            The Sone web interface
	 * @param requireLogin
	 *            Whether this page requires a login
	 */
	public SoneTemplatePage(String path, Template template, String pageTitleKey, WebInterface webInterface, boolean requireLogin) {
		super(path, webInterface.getTemplateContextFactory(), template, "noPermission.html");
		this.pageTitleKey = pageTitleKey;
		this.webInterface = webInterface;
		this.requireLogin = requireLogin;
	}

	//
	// PROTECTED METHODS
	//

	/**
	 * Returns the currently logged in Sone.
	 *
	 * @param toadletContext
	 *            The toadlet context
	 * @return The currently logged in Sone, or {@code null} if no Sone is
	 *         currently logged in
	 */
	protected Sone getCurrentSone(ToadletContext toadletContext) {
		return webInterface.getCurrentSoneCreatingSession(toadletContext);
	}

	protected Sone getCurrentSoneWithoutCreatingSession(ToadletContext toadletContext) {
		return webInterface.getCurrentSoneWithoutCreatingSession(toadletContext);
	}

	/**
	 * Sets the currently logged in Sone.
	 *
	 * @param toadletContext
	 *            The toadlet context
	 * @param sone
	 *            The Sone to set as currently logged in
	 */
	protected void setCurrentSone(ToadletContext toadletContext, Sone sone) {
		webInterface.setCurrentSone(toadletContext, sone);
	}

	//
	// TEMPLATEPAGE METHODS
	//

	/**
	 * {@inheritDoc}
	 */
	@Override
	protected String getPageTitle(FreenetRequest request) {
		if (pageTitleKey != null) {
			return webInterface.getL10n().getString(pageTitleKey);
		}
		return "";
	}

	/**
	 * {@inheritDoc}
	 */
	@Override
	protected List<Map<String, String>> getAdditionalLinkNodes(FreenetRequest request) {
		return ImmutableList.<Map<String, String>> builder().add(ImmutableMap.<String, String> builder().put("rel", "search").put("type", "application/opensearchdescription+xml").put("title", "Sone").put("href", "http://" + request.getHttpRequest().getHeader("host") + "/Sone/OpenSearch.xml").build()).build();
	}

	/**
	 * {@inheritDoc}
	 */
	@Override
	protected Collection<String> getStyleSheets() {
		return Arrays.asList("css/sone.css");
	}

	/**
	 * {@inheritDoc}
	 */
	@Override
	protected String getShortcutIcon() {
		return "images/icon.png";
	}

	/**
	 * Returns whether this page requires the user to log in.
	 *
	 * @return {@code true} if the user is required to be logged in to use this
	 *         page, {@code false} otherwise
	 */
	protected boolean requiresLogin() {
		return requireLogin;
	}

	/**
	 * {@inheritDoc}
	 */
	@Override
	protected final void processTemplate(FreenetRequest request, TemplateContext templateContext) throws RedirectException {
		super.processTemplate(request, templateContext);
		Sone currentSone = getCurrentSoneWithoutCreatingSession(request.getToadletContext());
		templateContext.set("core", webInterface.getCore());
		templateContext.set("currentSone", currentSone);
		templateContext.set("localSones", webInterface.getCore().getLocalSones());
		templateContext.set("request", request);
		templateContext.set("currentVersion", SonePlugin.getPluginVersion());
		templateContext.set("hasLatestVersion", webInterface.getCore().getUpdateChecker().hasLatestVersion());
		templateContext.set("latestEdition", webInterface.getCore().getUpdateChecker().getLatestEdition());
		templateContext.set("latestVersion", webInterface.getCore().getUpdateChecker().getLatestVersion());
		templateContext.set("latestVersionTime", webInterface.getCore().getUpdateChecker().getLatestVersionDate());
		List<Notification> notifications = new ArrayList<Notification>(webInterface.getNotifications(currentSone));
		Collections.sort(notifications, Notification.CREATED_TIME_SORTER);
		templateContext.set("notifications", notifications);
		templateContext.set("notificationHash", notifications.hashCode());
		handleRequest(request, templateContext);
	}

	protected void handleRequest(@Nonnull FreenetRequest request, @Nonnull TemplateContext templateContext) throws RedirectException {
	}

	/**
	 * {@inheritDoc}
	 */
	@Override
	protected String getRedirectTarget(FreenetRequest request) {
		if (requiresLogin() && (getCurrentSoneWithoutCreatingSession(request.getToadletContext()) == null)) {
			HTTPRequest httpRequest = request.getHttpRequest();
			String originalUrl = httpRequest.getPath();
			if (httpRequest.hasParameters()) {
				StringBuilder requestParameters = new StringBuilder();
				for (String parameterName : httpRequest.getParameterNames()) {
					if (requestParameters.length() > 0) {
						requestParameters.append("&");
					}
					String[] parameterValues = httpRequest.getMultipleParam(parameterName);
					for (String parameterValue : parameterValues) {
						requestParameters.append(urlEncode(parameterName)).append("=").append(urlEncode(parameterValue));
					}
				}
				originalUrl += "?" + requestParameters.toString();
			}
			return "login.html?target=" + urlEncode(originalUrl);
		}
		return null;
	}

	private static String urlEncode(String value) {
		try {
			return URLEncoder.encode(value, "UTF-8");
		} catch (UnsupportedEncodingException uee1) {
							/* A JVM without UTF-8? I don’t think so. */
			throw new RuntimeException(uee1);
		}
	}

	/**
	 * {@inheritDoc}
	 */
	@Override
	protected boolean isFullAccessOnly() {
		return webInterface.getCore().getPreferences().isRequireFullAccess();
	}

	/**
	 * {@inheritDoc}
	 */
	@Override
	public boolean isEnabled(ToadletContext toadletContext) {
		if (webInterface.getCore().getPreferences().isRequireFullAccess() && !toadletContext.isAllowedFullAccess()) {
			return false;
		}
		if (requiresLogin()) {
			return getCurrentSoneWithoutCreatingSession(toadletContext) != null;
		}
		return true;
	}

}

File: src/main/kotlin/net/pterodactylus/sone/web/pages/SoneTemplatePage.kt
package net.pterodactylus.sone.web.pages

import freenet.clients.http.*
import net.pterodactylus.sone.data.*
import net.pterodactylus.sone.main.*
import net.pterodactylus.sone.utils.*
import net.pterodactylus.sone.web.*
import net.pterodactylus.sone.web.page.*
import net.pterodactylus.util.notify.*
import net.pterodactylus.util.template.*
import net.pterodactylus.util.web.*
import java.net.*

/**
 * Base page for the Sone web interface.
 */
open class SoneTemplatePage @JvmOverloads constructor(
		path: String,
		private val webInterface: WebInterface,
		loaders: Loaders,
		templateRenderer: TemplateRenderer,
		private val pageTitleKey: String? = null,
		private val requiresLogin: Boolean = false,
		private val pageTitle: (FreenetRequest) -> String = { pageTitleKey?.let(webInterface.l10n::getString) ?: "" }
) : FreenetTemplatePage(path, templateRenderer, loaders, "noPermission.html") {

	private val core = webInterface.core
	private val sessionProvider: SessionProvider = webInterface

	protected fun getCurrentSone(toadletContext: ToadletContext, createSession: Boolean = true) =
			sessionProvider.getCurrentSone(toadletContext, createSession)

	protected fun setCurrentSone(toadletContext: ToadletContext, sone: Sone?) =
			sessionProvider.setCurrentSone(toadletContext, sone)

	fun requiresLogin() = requiresLogin

	override fun getPageTitle(request: FreenetRequest) = getPageTitle(request.toSoneRequest(core, webInterface))

	open fun getPageTitle(soneRequest: SoneRequest) = pageTitle(soneRequest)

	override val styleSheets = listOf("css/sone.css")

	override val shortcutIcon = "images/icon.png"

	override fun getAdditionalLinkNodes(request: FreenetRequest) =
			listOf(mapOf(
					"rel" to "search",
					"type" to "application/opensearchdescription+xml",
					"title" to "Sone",
					"href" to "http://${request.httpRequest.getHeader("host")}/Sone/OpenSearch.xml"
			))

	final override fun processTemplate(request: FreenetRequest, templateContext: TemplateContext) {
		super.processTemplate(request, templateContext)
		templateContext["preferences"] = core.preferences
		templateContext["currentSone"] = getCurrentSone(request.toadletContext)
		templateContext["localSones"] = core.localSones
		templateContext["request"] = request
		templateContext["currentVersion"] = SonePlugin.getPluginVersion()
		templateContext["hasLatestVersion"] = core.updateChecker.hasLatestVersion()
		templateContext["latestEdition"] = core.updateChecker.latestEdition
		templateContext["latestVersion"] = core.updateChecker.latestVersion
		templateContext["latestVersionTime"] = core.updateChecker.latestVersionDate
		webInterface.getNotifications(getCurrentSone(request.toadletContext)).sortedBy(Notification::getCreatedTime).run {
			templateContext["notifications"] = this
			templateContext["notificationHash"] = this.hashCode()
		}
		handleRequest(request, templateContext)
	}

	open fun handleRequest(freenetRequest: FreenetRequest, templateContext: TemplateContext) {
		handleRequest(freenetRequest.toSoneRequest(core, webInterface), templateContext)
	}

	open fun handleRequest(soneRequest: SoneRequest, templateContext: TemplateContext) {
	}

	override fun getRedirectTarget(request: FreenetRequest): String? {
		if (requiresLogin && getCurrentSone(request.toadletContext) == null) {
			val parameters = request.httpRequest.parameterNames
					.flatMap { name -> request.httpRequest.getMultipleParam(name).map { name to it } }
					.joinToString("&") { "${it.first.urlEncode}=${it.second.urlEncode}" }
					.emptyToNull
			return "login.html?target=${request.httpRequest.path}${parameters?.let { ("?" + it).urlEncode } ?: ""}"
		}
		return null
	}

	private val String.urlEncode: String get() = URLEncoder.encode(this, "UTF-8")

	override fun isEnabled(toadletContext: ToadletContext) =
			isEnabled(SoneRequest(toadletContext.uri, Method.GET, HTTPRequestImpl(toadletContext.uri, "GET"), toadletContext, webInterface.l10n, webInterface.sessionManager, core, webInterface))

	open fun isEnabled(soneRequest: SoneRequest) = when {
		requiresLogin && getCurrentSone(soneRequest.toadletContext) == null -> false
		core.preferences.requireFullAccess && !soneRequest.toadletContext.isAllowedFullAccess -> false
		else -> true
	}

}
